### PR TITLE
Process-specific working directory

### LINKED ISSUE #557: [RFC 0081] Process Specific Working Directories
[RFC 0081](https://github.com/buildpacks/rfcs/blob/main/text/0081-process-specific-working-directory.md)
https://github.com/buildpacks/rfcs/pull/144/files

### PR DESCRIPTION
**Pull request by mboldt**
Resolves #557

Thanks to @natalieparellano for review and feedback.

### CODE REVIEW

**Comment by mboldt** [comment]
Ack the Windows CI failure. That test passes in my Windows dev environment. Investigating.

**Comment by mboldt** [comment]
Found the Windows issue - the `cd` command needs the `/d` flag to change drives. On my dev machine, the working directory and the code were on the same drive, so it wasn't an issue. In CI, they are not. Will fix that up.

File: launch/launcher_test.go [hunk 1]
```diff
@@ -111,8 +116,9 @@ func testLauncher(t *testing.T, when spec.G, it spec.S) {
 
 		it.Before(func() {
 			process = launch.Process{
-				Command: "command",
-				Args:    []string{"arg1", "arg2"},
+				Command:          "command",
+				Args:             []string{"arg1", "arg2"},
+				WorkingDirectory: launcher.AppDir,
```
Comment (jabrown85) [inline launch/launcher_test.go:121]: I would feel better with a `WorkingDirectory: ""` test as well somewhere in here. To confirm we fallback to to `CNB_APP_DIR`

Comment (mboldt) [reply launch/launcher_test.go]: Added in 05d3979731525ab4a2f8acef7de83a8da4785861

**Comment by mboldt** [comment]
I don't know what is going on with the linux-arm64 build. It says it failed, but when I click "Details" it looks hung. Does someone have the capability to retry that/

**Comment by natalieparellano** [comment]
> I don't know what is going on with the linux-arm64 build.

It's probably the arm worker acting up again. I'll restart it...

**Comment by natalieparellano** [comment]
Green! 🎉 🎉 🎉  Taking a shot at user acceptance...

**Comment by natalieparellano** [comment]
Tested on Linux, it looks great! Looking at Windows...

bin/build:
```
#!/usr/bin/env bash
set -eo pipefail

echo "---> Hello processes buildpack"

# INPUT ARGUMENTS

env_dir=$2/env
layers_dir=$1
plan_path=$3

# ADD SYS-INFO LAYER DIR

echo "---> Adding sys-info layer dir"
sysinfo_layer_dir="${layers_dir}/sys-info"
mkdir -p ${sysinfo_layer_dir}
cat > "${sysinfo_layer_dir}.toml" << EOF
[types]
launch = true
EOF

# ADD SYS-INFO INDIRECT PROCESS

echo "---> Adding sys-info indirect process"
sysinfo_script="${sysinfo_layer_dir}/sys-info.sh"
cat > ${sysinfo_script} <<EOL
#!/usr/bin/env bash
echo "     work dir:" \$(pwd)
echo "     env vars:"
export | sed 's/^/       /'
EOL
chmod +x ${sysinfo_script}

cat >> "${layers_dir}/launch.toml" <<EOL
[[processes]]
type = "sys-info"
command = "${sysinfo_script}"
working-directory = "/cnb"
EOL

# ADD SYS-INFO INDIRECT PROCESS WITH ARGS

echo "---> Adding sys-info indirect process with args"
sysinfo_script="${sysinfo_layer_dir}/sys-info-args.sh"
cat > ${sysinfo_script} <<EOL
#!/usr/bin/env bash
echo "     first arg:" \$1
echo "     work dir:" \$(pwd)
echo "     env vars:"
export | sed 's/^/       /'
EOL
chmod +x ${sysinfo_script}

cat >> "${layers_dir}/launch.toml" <<EOL
[[processes]]
type = "sys-info-args"
command = "${sysinfo_script}"
args = ["some-arg"]
working-directory = "/cnb"
EOL


# ADD SYS-INFO DIRECT PROCESS

echo "---> Adding sys-info direct process"
sysinfo_script="${sysinfo_layer_dir}/sys-info-direct.sh"
cat > ${sysinfo_script} <<EOL
#!/usr/bin/env bash
echo "     work dir:" \$(pwd)
echo "     env vars:"
export | sed 's/^/       /'
EOL
chmod +x ${sysinfo_script}

cat >> "${layers_dir}/launch.toml" <<EOL
[[processes]]
type = "sys-info-direct"
command = "${sysinfo_script}"
direct = true
working-directory = "/cnb"
EOL

echo "---> Done"
```
```
$  pack inspect test-app
...
Processes:
  TYPE                   SHELL        COMMAND                                                          ARGS
  sys-info               bash         /layers/samples_hello-processes/sys-info/sys-info.sh             
  sys-info-args          bash         /layers/samples_hello-processes/sys-info/sys-info-args.sh        some-arg
  sys-info-direct                     /layers/samples_hello-processes/sys-info/sys-info-direct.sh
```
```
$ docker run --entrypoint sys-info-args test-app
$ docker run --entrypoint sys-info test-app
$ docker run --entrypoint sys-info-direct test-app
```
^^ all show `work dir: /cnb`

**Comment by natalieparellano** [comment]
Also confirmed that when buildpack api is < 0.8 we get the following warnings:
```
Warning: Warning: process working directory isn't supported in this buildpack api version. Ignoring working directory for process 'sys-info'
Warning: Warning: process working directory isn't supported in this buildpack api version. Ignoring working directory for process 'sys-info-args'
Warning: Warning: process working directory isn't supported in this buildpack api version. Ignoring working directory for process 'sys-info-direct'
```

**Review by jabrown85 — APPROVED** [review]

**Review by natalieparellano — APPROVED** [review]
Windows look good :) Thanks so much for the contribution @mboldt !

**Pull request merged**